### PR TITLE
Update balances onchain after tx confirm

### PR DESCRIPTION
**Pull request by brunobar79**
Pulled out `fetchAssetsBalancesAndPrices` into its own action with optional pricing info from coingecko and polling.

That way the fallbackExplorer can continue using it in the same way, but we now have the option to trigger a one time only onchain balance update  on demand.

Added the logic to do so at the App.js level after we listen to a tx confirmation event.

PoW:

QA: 
- Affected balance should be updated around 5 seconds after the transaction was confirmed in the app.
- Send any asset to the selected account (from outside the app) and pull to refresh any time after the tx was confirmed. The balance should be updated

PoW: https://recordit.co/JMHQIoVKBg

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -249,12 +252,24 @@ class App extends Component {
...
-    }, 10000);
+  handleTransactionConfirmed = tx => {
+    logger.log('reloading explorer data in 10');
+    const network = ethereumUtils.getNetworkFromChainId(tx.chainId);
+    const isL2 = isL2Network(network);
+    setTimeout(
+      () => {
+        if (isL2) {
+          logger.log('Reloading all data from L2 explorers!');
+          store.dispatch(explorerInitL2());
```
Comment (skylarbarrera) [inline src/App.js:263]: is it worth only reloading the specific L2 we need to check or nah?

File: src/App.js [hunk 2]
```diff
@@ -249,12 +252,24 @@ class App extends Component {
...
+        } else {
+          logger.log('fetching onchain balances NOW!');
+          store.dispatch(
+            fetchOnchainBalances({ keepPolling: false, withPrices: false })
+          );
+        }
+      },
+      isL2 ? 10000 : 5000
```
Comment (skylarbarrera) [inline src/App.js:271]: why longer for l2?

Comment (skylarbarrera) [reply src/App.js:271]: jk i saw waynebro's post

**Review by skylarbarrera — APPROVED** [review]
🔥🔥🔥

File: src/App.js (hunk 1 above)
Comment (brunobar79) [reply src/App.js:263]: nah

File: src/redux/fallbackExplorer.js [hunk 3]
```diff
@@ -286,60 +282,72 @@ export const fallbackExplorerInit = () => async (dispatch, getState) => {
...
+      )
+    );
+    lastUpdatePayload = newPayload;
+  }
 
+  if (keepPolling) {
     const fallbackExplorerBalancesHandle = setTimeout(
-      fetchAssetsBalancesAndPrices,
+      () =>
+        dispatch(dispatch(fetchOnchainBalances({ keepPolling, withPrices }))),
```
Comment (jinchung) [inline src/redux/fallbackExplorer.js:350]: `dispatch(dispatch(`

**Comment by jinchung** [comment]
line 356: `() => dispatch(findNewAssetsToWatch(accountAddress)),` doesn't need `accountAddress` passed along - I thought we had fixed this in a separate review a long time ago but maybe not

**Comment by jinchung** [comment]
`findAssetsToWatch` on line 51: is it ok that it doesn't pass along coingecko ids? is this expected behavior? (`discoverTokens` that is called doesn't seem to do checks on coingecko ids being non null before accessing it)

File: src/redux/fallbackExplorer.js [hunk 4]
```diff
@@ -357,8 +365,34 @@ export const fallbackExplorerInit = () => async (dispatch, getState) => {
...
+    const newMainnetAssets = await findAssetsToWatch(
+      accountAddress,
+      latestTxBlockNumber,
+      dispatch,
+      coingeckoIds
+    );
+
+    await dispatch({
+      payload: {
+        mainnetAssets: mainnetAssets.concat(newMainnetAssets),
```
Comment (jinchung) [inline src/redux/fallbackExplorer.js:389]: this existed already but is this right to just do a straight concat of the mainnetAssets with newMainnetAssets without any deduping / uniqueBy?

Comment (jinchung) [inline src/redux/fallbackExplorer.js:389]: this existed already but is this right to just do a straight concat of the mainnetAssets with newMainnetAssets without any deduping / uniqueBy? deduping logic exists elsewhere but i don't see it here

**Comment by jinchung** [comment]
when adding/switching to a new wallet, `latestTxBlockNumber` does not get reset (causing the next call to `getTokenTxDataFromEtherscan` to come back empty

**Comment by jinchung** [comment]
we should use an ETHERSCAN_API_KEY for these calls as I often see rate limits (with an API key, it is 5/sec/IP for valid API key, vs just 1 every 5 seconds without API key)

File: src/App.js (hunk 1 above)
Comment (jinchung) [reply src/App.js:263]: skylar i like ur thinking tho

🤨 bruno over here acting like we rich in compute units 😛

File: src/redux/fallbackExplorer.js (hunk 4 above)
Comment (brunobar79) [reply src/redux/fallbackExplorer.js:389]: It happens on line 63

**Comment by brunobar79** [comment]
> when adding/switching to a new wallet, `latestTxBlockNumber` does not get reset (causing the next call to `getTokenTxDataFromEtherscan` to come back empty

Now added a call to `fallbackExplorerClearState` inside `explorerClearState` so we're good

File: src/redux/explorer.js [hunk 5]
```diff
@@ -246,7 +247,12 @@ const isValidAssetsResponseFromZerion = msg => {
   return false;
 };
 
-export const explorerClearState = () => dispatch => {
+export const explorerClearState = () => (dispatch, getState) => {
+  const { fallback } = getState().explorer;
+  if (fallback) {
+    logger.log('😬 Disabling fallback data provider!');
+    dispatch(fallbackExplorerClearState());
```
Comment (jinchung) [inline src/redux/explorer.js:254]: `fallbackExplorerClearState` is already done as part of `disableFallbackIfNeeded` in line 256

Comment (brunobar79) [reply src/redux/explorer.js]: oops

File: src/handlers/covalent.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  COVALENT_ANDROID_API_KEY,
+  // @ts-ignore
+  COVALENT_IOS_API_KEY,
+} from 'react-native-dotenv';
+import { rainbowFetch } from '../rainbow-fetch';
+import { EthereumAddress } from '@rainbow-me/entities';
+import Logger from 'logger';
+export const getAssetsFromCovalent = async (
+  chainId: Number,
+  address: EthereumAddress,
```
Comment (jinchung) [inline src/handlers/covalent.ts:13]: can we stick to `accountAddress` naming convention

Comment (jinchung) [reply src/handlers/covalent.ts:13]: same thing in the files where this is called

Comment (jinchung) [reply src/handlers/covalent.ts:13]: also while we're at it, can you insert a space between lines 10 and 11

Comment (brunobar79) [reply src/handlers/covalent.ts:13]: Sure. we're moving it to the backend anyways so I don't care.

File: src/redux/fallbackExplorer.js [hunk 7]
```diff
@@ -221,51 +282,65 @@ const fetchAssetBalances = async (tokens, address, network) => {
...
+    allAssets,
+    genericAssets
+  );
 
-    await dispatch({
+  const { assets: accountAssets } = getState().data;
+  let assets =
+    network === NetworkTypes.mainnet
+      ? uniqBy(
+          [...mainnetAssets, ...covalentMainnetAssets],
```
Comment (jinchung) [inline src/redux/fallbackExplorer.js:310]: `covalentMainnetAssets` can be null so we should do a check before doing destructuring

**Review by jinchung — APPROVED** [review]
🌮

**Review by ibrahimtaveras00 — APPROVED** [review]
looks good on mainnet on both OS's with ETH and ERC20 tokens 
QA Passed 👍🏽

**Pull request merged**